### PR TITLE
Payment, Reviews Booking

### PR DESCRIPTION
**Pull request by CamiloGilFranco**
## SUMMARY

- Service and controller are created to obtain a user with their payment details

- Authorization to create a new user address is added

- A new controller and service route is created for the review count of a restaurant

- A new controller and service route is created for the restaurant rating update

- Restriction by authorization to create a comment is added

- A new controller and service route is created for the restriction of multiple comments from a single user in the same restaurant.

- Modify the controller and createReserve service to apply authorization

- Route and controller are created for the payment gateway

- The "quantity" column is added to the order_details entity

- The order_details seeder is modified to generate data from the new column

- A relationship is created between the entity user_addresses and order

- The order seeder is modified to add the delivery address to each order

- Route, controller and service are created to add several order_details records in a single request

- The CreateOrder controller and service are modified to verify user authorization

- The relationship entity diagram is updated with the new changes implemented

### CODE REVIEW

File: src/api/restaurants/restaurants.controller.ts [hunk 1]
```diff
@@ -117,3 +118,17 @@ export const deleteRestaurantController = async (
...
+
+//update restaurant rating
+export const updateRestaurantRatingController = async (
+  req: Request,
+  res: Response
+) => {
+  try {
+    const { id_restaurant, rating } = req.body;
+    await updateRestaurantRating(id_restaurant, rating);
+    res.status(200).json({ message: "Restaurant Raring updated" });
```
Comment (ht1204) [inline src/api/restaurants/restaurants.controller.ts:130]: `Restaurant Rating updated`

File: src/api/reservations/reservation.controller.ts [hunk 2]
```diff
@@ -38,12 +39,19 @@ export const getByIdReservationController = async (
...
 export const createReservationController = async (
-  req: Request,
+  req: Request & AuthUser,
   res: Response,
   next: NextFunction
 ) => {
   try {
-    const reservation = await createReservation(req.body);
+    const { user } = req;
+    const { id_restaurant, id_venue, date } = req.body;
```
Comment (ht1204) [inline src/api/reservations/reservation.controller.ts:48]:
```js
const {
  user,
  body: {
   id_restaurant,
   id_venue,
  date
  }
} = req;
```

File: src/api/reservations/reservation.service.ts [hunk 3]
```diff
@@ -15,18 +15,12 @@ export const getByIdReservation = (id: string) => {
...
   return prisma.reservations.create({
     data: {
-      date_hour,
-      restaurantsId_restaurant,
-      restaurant_venuesId_restaurant_venue,
-      usersUser_id,
+      date_hour: input.date,
+      restaurantsId_restaurant: input.id_restaurant,
+      restaurant_venuesId_restaurant_venue: input.id_venue,
+      usersUser_id: input.user,
```
Comment (ht1204) [inline src/api/reservations/reservation.service.ts:23]: destructuring could be an enhancement.

File: src/api/reviews/reviews.controller.ts [hunk 4]
```diff
@@ -5,21 +5,28 @@ import {
...
-  res: Response,
-  next: NextFunction
+export const getAllReviewsRestaurantController = async (
+  req: Request & AuthUser,
+  res: Response
 ) => {
   try {
-    const reviews = await getAllReviews();
+    const { user } = req;
+    const { id_restaurant } = req.query;
```
Comment (ht1204) [inline src/api/reviews/reviews.controller.ts:19]: `const { user, query: { id_restaurant }} = req;`

File: src/api/reviews/reviews.controller.ts [hunk 5]
```diff
@@ -42,15 +49,37 @@ export const getReviewByIdController = async (
...
-  req: Request,
-  res: Response,
-  next: NextFunction
+  req: Request & AuthUser,
+  res: Response
 ) => {
   try {
-    const review = await createReview(req.body);
+    const { user } = req;
+    const { id_restaurant, title, rating, comment } = req.body;
```
Comment (ht1204) [inline src/api/reviews/reviews.controller.ts:57]: destructuring for coding enhancement.

File: src/api/restaurants/restaurants.services.ts [hunk 6]
```diff
@@ -165,3 +165,16 @@ export const deleteRestaurant = (id: string) => {
     },
   });
 };
+
+//update restaurant rating
+
+export const updateRestaurantRating = (id_restaurant: any, rating: any) => {
```
Comment (ht1204) [inline src/api/restaurants/restaurants.services.ts:171]: why not datatype for params?

File: src/api/reviews/reviews.controller.ts [hunk 7]
```diff
@@ -42,15 +49,37 @@ export const getReviewByIdController = async (
...
+      user,
+      id_restaurant,
+      title,
+      rating: parseInt(rating),
+      comment,
+    });
     res.status(201).json({ message: "Review created", data: review });
   } catch (error) {
-    next(error);
+    res.status(400).json({ message: "Error...." });
```
Comment (ht1204) [inline src/api/reviews/reviews.controller.ts:68]: An error message more clear here, please.

File: src/api/restaurants/restaurants.services.ts (hunk 6 above)
Comment (CamiloGilFranco) [reply src/api/restaurants/restaurants.services.ts:171]: I am still not very clear about the typing, I would like us to meet one day to understand this issue and clean up the code

File: src/api/reservations/reservation.service.ts (hunk 3 above)
Comment (CamiloGilFranco) [reply src/api/reservations/reservation.service.ts:23]: I tried to write it like this but doesn't work
const {
  date_hour: date,
  restaurantsId_restaurant: id_restaurant,
  restaurant_venuesId_restaurant_venue: id_venue,
  usersUser_id: user
} = input;

Comment (ht1204) [reply src/api/reservations/reservation.service.ts:23]: ok, no problem, thx 4 updating.

**Review by ht1204 — APPROVED** [review]

**Pull request merged**